src/transformers/modeling_attn_mask_utils.py

```diff
@@ -234,6 +239,56 @@ def _unmask_unattended(
 
         return expanded_mask.mul(~torch.all(expanded_mask == min_dtype, dim=-1, keepdim=True))
 
+    @staticmethod
+    def _ignore_causal_mask_sdpa(
+        attention_mask: Optional[torch.Tensor],
+        inputs_embeds: torch.Tensor,
+        past_key_values_length: int,
+    ) -> Tuple[torch.Tensor, bool]:
+        """
+        In case no token is masked in the `attention_mask` argument, we simply set it to `None` for the cases `query_length == 1` and
+        `key_value_length == query_length`, and rely instead on SDPA `is_causal` argument to use causal/non-causal masks,
+        allowing to dispatch to the flash attention kernel (that can otherwise not be used if a custom `attn_mask` is passed).
+        """
+
+        batch_size, query_length = inputs_embeds.shape[0], inputs_embeds.shape[1]
+        key_value_length = query_length + past_key_values_length
+
+        # torch.jit.trace, symbolic_trace and torchdynamo with fullgraph=True are unable to capture the controlflow `is_causal=attention_mask is None and q_len > 1`
+        # used as an SDPA argument. We keep compatibility with these tracing tools by always using SDPA's `attn_mask` argument in case we are tracing.
+        # TODO: For dynamo, rather use a check on fullgraph=True once this is possible (https://github.com/pytorch/pytorch/pull/120400).
+        is_tracing = (
+            torch.jit.is_tracing()
+            or isinstance(inputs_embeds, torch.fx.Proxy)
+            or (hasattr(torch, "_dynamo") and torch._dynamo.is_compiling())
+        )
+
+        mask_none_and_is_tracing = False
+        if attention_mask is not None:
+            if len(attention_mask.shape) == 4:
+                expected_shape = (batch_size, 1, query_length, key_value_length)
+                if tuple(attention_mask.shape) != expected_shape:
+                    raise ValueError(
+                        f"Incorrect 4D attention_mask shape: {tuple(attention_mask.shape)}; expected: {expected_shape}."
+                    )
+            elif not is_tracing and torch.all(attention_mask == 1):
+                if query_length == 1:
+                    # For query_length == 1, causal attention and bi-directional attention are the same.
+                    attention_mask = None
+                elif key_value_length == query_length:
+                    attention_mask = None
+                else:
+                    # Unfortunately, for query_length > 1 and key_value_length != query_length, we cannot generally ignore the attention mask, as SDPA causal mask generation
+                    # may be wrong. We will set `is_causal=False` in SDPA and rely on Transformers attention_mask instead, hence not setting it to None here.
+                    # Reference: https://github.com/pytorch/pytorch/issues/108108
+                    pass
```

Comment:
> It appears that this issue is resolved in PyTorch 2.3 RC in https://github.com/pytorch/pytorch/pull/114823. Although supporting the new API might be best left to a future PR.

Thank you, it might be worth to revisit in an other PR.